node/coinstacks/solana/api/src/websocket.ts

```diff
@@ -0,0 +1,117 @@
+import { Logger } from '@shapeshiftoss/logger'
+import { WebsocketClient as BaseWebsocketClient, Args, Options, Subscription } from '@shapeshiftoss/websocket'
+import WebSocket from 'ws'
+import {
+  isWebsocketResponse,
+  isWebsocketSubscribeResponse,
+  WebsocketResponse,
+  WebsocketSubscribeResponse,
+} from './types'
+import { Logs } from '@solana/web3.js'
+
+const logger = new Logger({ namespace: ['unchained', 'solana', 'websocket'], level: process.env.LOG_LEVEL })
+
+interface WebsocketArgs extends Omit<Args, 'logger'> {
+  transactionHandler: TransactionHandler | Array<TransactionHandler>
+}
+
+type TransactionHandler = (data: Logs) => Promise<void>
+
+export class WebsocketClient extends BaseWebsocketClient {
+  private handleTransaction: TransactionHandler | Array<TransactionHandler>
+  private addresses: Array<string> = []
+  private subscriptionsIds: Map<number, number> = new Map()
+  private currentSubscriptionId: number = 0
+
+  constructor(_url: string, args: WebsocketArgs, opts?: Options) {
+    const url = args.apiKey ? `${_url}?api-key=${args.apiKey}` : _url
+    super(url, { logger }, opts)
+    this.handleTransaction = args.transactionHandler
+    this.initialize()
+    this.socket.onmessage = (msg) => this.onMessage(msg)
+    this.socket.onopen = () => this.onOpen()
+  }
+
+  protected onOpen(): void {
+    super._onOpen()
+    if (this.addresses.length > 0) {
+      this.subscribeAddresses(this.addresses)
+    }
+  }
+
+  protected async onMessage(message: WebSocket.MessageEvent): Promise<void> {
+    try {
+      const res: WebsocketResponse | WebsocketSubscribeResponse = JSON.parse(message.data.toString())
+      if (isWebsocketSubscribeResponse(res)) {
+        this.subscriptionsIds.set(this.currentSubscriptionId, res.result)
```

Comment:
one more edge case I am thinking, by using this.currentSubscriptionId here, we don't actually know if this message was from the corresponding subscription. I was thinking we would actually use the `currentSubscriptionId` as the `id` in the subscribe message so we can look at the response and check if the `id` matches the `currentSubcriptionId` and if it doesn't, we can immediately unsubscribe to that subscription.

Reply:
Done in [37ba9ed](https://github.com/shapeshift/unchained/pull/1053/commits/37ba9edb1be37c7338ff96166dc24614031fbd65)